[WIP]Merge desktop code into develop branch

vkjr opened the pull request:
fixes #5083

### Summary:

This idea behind this PR is to merge mobile and desktop code and have it in the same branch. `desktop_develop_merge` branch contains code from `desktop` rebased on `develop`.

In most files code difference handled by `(if platform/desktop? ...)` statements. 

`vector_icons.cljs` works quite differently for desktop. It doesn't use `react-native-svg` library and able to display svg directly in `Image` component (thanks to underlying Qt).

`js_dependencies.cljs` also has a lot of changes between desktop and mobile versions. Mostly because not all 3rdparty dependencies are implemented and needed by desktop build.

The most problematic are files: `VERSION`, `package.json` and `package-lock.json`. They are different for mobile and desktop and `(if desktop?)` doesn't work there...
I saved both versions of these files - with `*.mobile` and `*.desktop` extensions.
If you want to build code of mobile app, you should invoke `scripts/prepare-for-platform.sh mobile`. `scripts/prepare-for-platform.sh desktop` for desktop build.
That script will replace `package.json`, `VERSION`, `.re-natal`, `package-lock.json` with platform-dependent versions of file, also it will remove `node_modules` directory.

I strongly dislike this approach but didn't come up with anything better so far. Would be glad to hear your propositions on how to make that better.



### Review notes (optional):
Here are steps to run `ios` app from the code:

```
scripts/prepare-for-platform.sh mobile
npm install
mvn -f modules/react-native-status/ios/RCTStatus dependency:unpack
ln -sf './node_modules/re-natal/index.js' './re-natal'
(In terminal 1:) make startdev-ios-simulator
(In terminal 2:) react-native run-ios
```

Here are steps to run `desktop` app from the code:

```
scripts/prepare-for-platform.sh desktop
npm install;
ln -sf './node_modules/re-natal/index.js' './re-natal'
react-native desktop; (answer `n` to any proposed replacing)
./re-natal use-figwheel;
./re-natal enable-source-maps;
(In terminal 1:) lein figwheel-repl desktop (wait until compilation finished)
(In terminal 2:) npm start
(In terminal 3:) ./ubuntu-server.js
(In terminal 3:) react-native run-desktop
```


### Testing notes (optional):


### Steps to test:
- Open mobile or desktop Status
- Check that it works correctly
- Step 3, etc.

<!-- (PRs will only be accepted if squashed into single commit.) -->

status: ready

vkjr committed:
UI works with react-native-desktop

vkjr committed:
Added prod build to status-react desktop

vkjr committed:
main.cpp added to git

vkjr committed:
Added changes to main.cpp that allows mac bundle building

vkjr committed:
Added script for making mac bundle

vkjr committed:
Fixed review notes

vkjr committed:
Insert newline by shift+enter (#4444)

* New line can be added in message by shift+enter. Account key copied to clipboard when pressed

* Fixed reiew note: removed unneccessary do

maxhora committed:
Desktop Jenkins ci build #4352

vkjr committed:
Figwheel fixed for desktop. Added few attempts to retrieve file content from bundler (#4599)

maxhora committed:
Linux Jenkins parallel build #4352

Commit:
Add DevEnvSetup and Troubleshooting docs for desktop app

Commit:
Clarify what 'node' means

Commit:
Fix chat list item reordering

Commit:
include version number on profile screen, read from local file

use i18n label for "Log out"

vkjr committed:
Desktop chat design implemented (#4791)

maxhora committed:
react-native version update (#4840)

* react-native version update
* Update react version to match react-native dependencies

vkjr committed:
Merged with develop branch

Commit:
Add share-contact-code

Qr code copy window

Copy code on-press handler

Add profile styles

Alignments, obsolete style removed

Add qr code close icon

Fix typo

Add copy-to-clipboard event handler and fx

Remove obsolete current-account binding

Formatting

Remove unneeded horizontal lines

Center qr code text

Center 'Share my contact code' button

Add green 'Copied to clipboard' tooltip view

Commit:
Fix key uniqueness in message history;
correct typo in Figwheel output
file name

Commit:
Fix 'Every element in a seq should have a unique :key' warning

maxhora committed:
Abort git rebase on Jenkins failure

maxhora committed:
Resolve Jenkins build and merge issues

Commit:
don't logout on version # click

Commit:
Stop node before starting on desktop platform

desktop? is not a function

Commit:
Always scroll chat view to the bottom

Commit:
Formatting

maxhora committed:
Use status-go from develop branch #4996

maxhora committed:
Cleanup disk space after build

maxhora committed:
Build macos dmg from locally copied official qt distribution

maxhora committed:
Preset keychain username on desktop

maxhora committed:
Clean build files

vkjr committed:
Jenkins now downloads package with correct Info.plist

vkjr committed:
Add execution permission to ubuntu-server

Commit:
Add fading for qr code copied tooltip

Commit:
Always clear intervals

Commit:
Change profile view styling

Commit:
Refresh right pane when navigating to a different tab

Commit:
Increase left pane width to 340

vkjr committed:
Saved files that are unique for develop branch

vkjr committed:
Fixed errors during rebase merge

vkjr committed:
More rebase errors fixed

vkjr committed:
Fixed to run the code on desktop

vkjr committed:
Fixed mobile version after merge with desktop branch

vkjr committed:
Merged with latest develop

vkjr committed:
Merge branch 'develop' into desktop_develop_merge

vkjr committed:
Set mobile version of files

vkjr committed:
Fixed few endlines

cammellos commented on .env:
```diff
@@ -17,4 +17,4 @@ DEBUG_WEBVIEW=1
 USE_SYM_KEY=0
-SPAM_BUTTON_DETECTION_ENABLED=1
\ No newline at end of file
+SPAM_BUTTON_DETECTION_ENABLED=0
```
Any reason this has been disabled?

vkjr replied:
No reason, will fix that.

cammellos commented on src/status_im/utils/config.cljs:
```diff
@@ -24,7 +24,7 @@
 (def universal-links-enabled? (enabled? (get-config :UNIVERSAL_LINK_ENABLED "1")))
 (def log-level
-  (-> (get-config :LOG_LEVEL "error")
+  (-> (get-config :LOG_LEVEL "debug")
```
Did you mean to change the loglevel or it was used locally?

vkjr replied:
Used locally, will fix that.

cammellos commented on src/status_im/ui/screens/events.cljs:
```diff
@@ -370,7 +374,7 @@
                          [:update-wallet]
                          [:update-transactions]
-                         [:get-fcm-token]
+                         (if-not platform/desktop? [:get-fcm-token])
```
I'd probably rather have an explicit method for this `platform/mobile?` rather then checking that is not desktop directly (the implementation might very well be `not desktop?`)

cammellos commented:
@vkjr 
about building for different platforms, one way could be to build desktop inside `docker`, we know we can do that for android ( I don't think it's possible for `ios` as it can't be build in a `linux` environment).

In that way you can have a different `Dockerfile` and it will not touch any of the local files: `ADD desktop/VERSION ./VERSION` etc

But you would be need to be able to build a `MACOS` binary inside a linux container, is that possible? otherwise it would not work.

`VERSION` we can probably fix using compile time environment variables, but `package.json` etc are still problematic

dmitryn commented:
Not sure it was discussed before, but what code does desktop and mobile share? 

Could that code be moved to separate repo (say `core`) and then used by git submodule? In the end, it will be 2 separate projects `status-mobile` and `status-desktop` each one having it's own dependencies in package.json and it's own build tools/scripts.

vkjr committed:
Fixed review notes from cammellos

vkjr commented:
@dmitryn, the desktop has own UI and shares all other functionality with mobile. Currently, it is all about messaging, wallet and dapps will be added later.
Probably splitting the projects is a good step to discuss after we have all code placed together. And that probably should be done by someone more experienced in mobile code than I :)

vkjr committed:
Single .re-natal now used for both mobile and desktop platform

vkjr commented:
I just managed to use one `.re-natal` file for both platform and desktop so one problem less.

cammellos commented:
@vkjr we can also simplify a bit `package.json` management by having 2 directories `desktop/package.json`  & `mobile/package.json`, and then you can run `npm install --prefix desktop/` which will create a `node_modules` directory in `desktop/node_modules` .

We can `symlink` `package.json`/`package-lock.json`/`node_modules` in the build/development scripts, and add these to gitignore. Just throwing around ideas :)

rcullito commented:
the directory approach to managing multiple package.jsons (`desktop/package.json` and mobile/package.json`) seems good. there might be some tooling we can experiment with here: https://yarnpkg.com/blog/2017/08/02/introducing-workspaces/

> Yarn Workspaces is a feature that allows users to install dependencies from multiple package.json files in subfolders of a single root package.json file, all in one go.

Open to comment with those more experienced w/ mobile dev on whether introducing yarn versus npm here introduces any other issues to be aware of. 

Nice job on getting the merge kicked off @vkjr!

<blockquote><img src="https://yarnpkg.com/assets/og_image.png" width="48" align="right"><div><img src="/favicon.ico" height="14"> Yarn</div><div><strong><a href="https://yarnpkg.com/blog/2017/08/02/introducing-workspaces/">Workspaces in Yarn</a></strong></div><div>Projects tend to grow over time, and, occasionally, some pieces of a project can be useful elsewhere in other projects. For example, Jest, being a generic te...</div></blockquote>

rcullito commented on react-native/src/status_im/react_native/js_dependencies.cljs:
```diff
@@ -1,31 +1,55 @@
+    (= (.-OS (.-Platform react-native)) "desktop"))
+
+(if desktopPlatform?
+  (do (def i18n                   (js/require "react-native-i18n"))
```
It feels like it would be clearer to split up desktop and mobile `defs` here into separate files and bring them into this namespace via `: refer :all` rather than having `do` perform these definitions via side effects.

vkjr replied:
@rcullito, thanks, I've created new PR for these changes, will fix your note there - #5203

vkjr committed:
Now using symlinks instead of FILENAME.desktop/FILENAME.mobile for platform-dependent files

vkjr commented:
@cammellos, good idea with symlinks. Implemented it. I made 2 folders with platform-dependent files for desktop and mobile. Unfortunately symlinking `node_modules` folder leads to errors when invoking `.re-natal enable-source-maps` and `npm start` also starts in a wrong folder.
So I modified `scripts/prepare_for_platform.sh` to remove `node_modules` and create appropriate symlinks.
Could you please check this changes?

@rcullito, thank you for pointing to yarn workspaces. If I understood correctly they are more usable when you have 2 projects in different folders that share some common code. Not so obvious for me how to use them right now, when we have one project. But they could be helpful if later we decide to split projects as @dmitryn suggested.

vkjr committed:
Merged with latest develop

vkjr commented:
This PR contains too many commit and can't be built by jenkins (since it tries to do a rebase).
So I recreated a PR and squashed commits. Please check - #5203. Jenkins built it correctly for ios/android.
This one will be closed now.
